File: system/multitier/backend-typescript/src/core/services/external/clock.gateway.ts
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { ClockGetTimeResponse } from '../../dtos/external/clock-get-time-response.dto';
import { errorMessage, fetchJson } from './fetch-json';

type ExternalSystemMode = 'real' | 'stub';

const isExternalSystemMode = (value: string): value is ExternalSystemMode =>
  value === 'real' || value === 'stub';

@Injectable()
export class ClockGateway {
  private readonly externalSystemMode: ExternalSystemMode;
  private readonly clockUrl: string;

  constructor(private readonly configService: ConfigService) {
    const externalSystemMode = this.configService.get<string>(
      'EXTERNAL_SYSTEM_MODE',
      'real',
    );
    if (!isExternalSystemMode(externalSystemMode)) {
      throw new Error(`Unknown external system mode: ${externalSystemMode}`);
    }
    this.externalSystemMode = externalSystemMode;
    this.clockUrl = this.configService.get<string>(
      'CLOCK_API_URL',
      'http://localhost:9001/clock',
    );
  }

  async getCurrentTime(): Promise<Date> {
    if (this.externalSystemMode === 'real') {
      return new Date();
    }
    return this.getStubTime();
  }

  private async getStubTime(): Promise<Date> {
    const url = `${this.clockUrl}/api/time`;
    try {
      const clockResponse = await fetchJson(url, ClockGetTimeResponse);
      return new Date(clockResponse.time);
    } catch (e) {
      throw new Error(`Failed to fetch current time. ${errorMessage(e)}`, {
        cause: e,
      });
    }
  }
}

File: system/multitier/backend-java/src/main/java/com/mycompany/myshop/backend/core/services/external/ClockGateway.java
package com.mycompany.myshop.backend.core.services.external;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import com.mycompany.myshop.backend.core.dtos.external.ClockGetTimeResponse;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Instant;

@Service
public class ClockGateway {

    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper()
            .registerModule(new JavaTimeModule());

    private final String externalSystemMode;
    private final String clockUrl;

    public ClockGateway(
            @Value("${external.system-mode}") String externalSystemMode,
            @Value("${clock.url}") String clockUrl) {
        this.externalSystemMode = externalSystemMode;
        this.clockUrl = clockUrl;
    }

    public Instant getCurrentTime() {
        if ("real".equals(externalSystemMode)) {
            return Instant.now();
        } else if ("stub".equals(externalSystemMode)) {
            return getStubTime();
        } else {
            throw new IllegalStateException("Unknown external system mode: " + externalSystemMode);
        }
    }

    private Instant getStubTime() {
        try {
            var httpClient = HttpClient.newBuilder()
                    .connectTimeout(java.time.Duration.ofSeconds(10))
                    .build();

            var url = clockUrl + "/api/time";
            var request = HttpRequest.newBuilder()
                    .uri(URI.create(url))
                    .timeout(java.time.Duration.ofSeconds(10))
                    .GET()
                    .build();

            var response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() != 200) {
                throw new IllegalStateException("Clock API returned status " + response.statusCode() +
                        ". URL: " + url + ". Response: " + response.body());
            }

            var clockResponse = OBJECT_MAPPER.readValue(response.body(), ClockGetTimeResponse.class);
            return clockResponse.getTime();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException("Failed to fetch current time from URL: " + clockUrl +
                    ". Error: " + e.getClass().getSimpleName() + ": " + e.getMessage(), e);
        } catch (Exception e) {
            throw new IllegalStateException("Failed to fetch current time from URL: " + clockUrl +
                    ". Error: " + e.getClass().getSimpleName() + ": " + e.getMessage(), e);
        }
    }
}
